src/hotspot/share/opto/subnode.cpp

```diff
@@ -235,8 +235,8 @@ Node *SubINode::Ideal(PhaseGVN *phase, bool can_reshape){
     return new SubINode(phase->intcon(0), in2->in(1));
   }
 
-  // Convert "0 - (x-y)" into "y-x"
-  if( t1 == TypeInt::ZERO && op2 == Op_SubI )
+  // Convert "0 - (x-y)" into "y-x", leave the double negation "-(-y)" to SubNode::Identity().
+  if(t1 == TypeInt::ZERO && op2 == Op_SubI && phase->type(in2->in(1)) != TypeInt::ZERO)
```

Comment:
Since you are modifying that code, please also add parentheses around the body, update the code style and add a whitespace between `if` and the condition.

Reply:
Thanks for your review, it's fixed.